README.md

```diff
@@ -4,227 +4,200 @@
 
-For more information on accepted commands for testing with Rake you can execute:
+An example of an ENV file can be found at `openshift/.env.test`
 
-> `$ bundle exec rake -T`
+### Automatic execution (with Makefile)
 
-### Testing API users from the outside
+Makefile rules can be run to execute the different Apisonator components
+with some predefined behaviour. To do this a file named `.env` in
+the `openshift` directory must be created before.
 
-You can test users of Apisonator API services by starting its dependencies and
-then apisonator itself:
+Once this has been performed, go to the `openshift` directory and execute
+one of the available Makefile commands to run Apisonator components:
 
-> `$ script/test_external`
+#### Apisonator Listener
 
-This will launch the application and wait for requests. You could now ie. launch
-`3scale_core`'s testing against this backend instance.
+Execute the Apisonator Listener, exposing the port 3001:
 
-You can hit CTRL+C at any time to exit. If you wanted to pass extra parameters
-for the launcher, such as daemonizing, you could run this:
+```
+make listener
+```
 
-> `$ script/test_external -- -d`
+#### Apisonator Worker
 
-## Running
+```
+make worker
+```
 
-Apisonator has its own application runner for the listener service:
+#### Apisonator failed jobs rescheduler
 
-`bundle exec 3scale_backend help`
+Execute the 'cron' Apisonator component:
 
-That will give you a help message so that you figure out how to invoke it. Some
-advanced usage patterns are available, as well as support for multiple
-application servers. The usual invocation would look something like:
+```
+make cron
+```
 
-> `$ bundle exec 3scale_backend start -p 3001`
+#### Apisonator bash shell
 
-That would start the listener service in port 3001. There are however additional
-commands and flags that are worth knowing, so please take a look at the help
-message. An interesting command would be this one:
+Execute a bash shell with the Apisonator source code with all the available
+components:
 
-> `$ bundle exec 3scale_backend -s puma -X "-w 3" start -p 3001`
+```
+make bash
+```
 
-That would be identical to the command above, except it would force the server
-to be Puma with 3 workers. You can obtain more info about the server specific
-options with:
+### Manual execution
 
-> `$ bundle exec 3scale_backend -s puma help-server`
+Another way of executing the Apisonator components is by running a container
+using the previously generated Apisonator image:
 
-Additional information which is really important to how the application runs can
-be found in the manifest:
+#### Apisonator Listener
 
-> `$ bundle exec 3scale_backend manifest`
+To run an Apisonator listener, the script bin/3scale_backend is used. To
+run it from a previously generated Apisonator docker image:
 
-## Documentation
+```
+docker run -it amp:apisonator-<version_number> -p 3001:3001 --env-file <myenv_file> 3scale_backend start -p 3001 -x /dev/stdout
```

Comment:
hm, we have an `EXPOSE 3000` command in the Dockerfile, maybe we can take advantage of that (plus it's also the default 3scale_backend listener port) to shorten this command here (ie. move to 3000 and we lose the -p 3001 in the start command, and maybe we can shorten the docker `-p 3000:3000` or use `--expose 3000`?

Reply:
As far as I know, the -p parameter and the --expose parameter do different things. The expose parameter does not really publish the ports:
https://docs.docker.com/engine/reference/builder/
```
The EXPOSE instruction does not actually publish the port. It functions as a type of documentation between the person who builds the image and the person who runs the container, about which ports are intended to be published. To actually publish the port when running the container, use the -p flag on docker run to publish and map one or more ports, or the -P flag to publish all exposed ports and map them to high-order ports.
```
Regarding the port number, although in the Dockerfile it exports port 3000 I chose 3001 because the automatic Makefile rules start the backend listener in port 3001 (see file openshift/mk/backend.mk) and I decided to be consistent with that because we have the othersection with the execution via Makefile

Reply:
ok